provider/datagen/src/collator/transform.rs

```diff
@@ -0,0 +1,284 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_codepointtrie::CodePointTrie;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::path::Path;
+use zerovec::ZeroVec;
+use std::str::FromStr;
+
+use crate::DatagenOptions;
+use crate::uprops::codepointtrie::TomlCodePointTrie;
+use crate::uprops::reader::*;
+use eyre::WrapErr;
+use icu_locid::LanguageIdentifier;
+use icu_provider::datagen::IterableResourceProvider;
+use icu_provider::prelude::*;
+use writeable::Writeable;
+use icu_locid::Locale;
+use icu_locid::unicode_ext_key;
+use icu_locid::extensions::unicode::Value;
+
+#[derive(serde::Deserialize)]
+pub struct CollationData {
+    pub trie: TomlCodePointTrie,
+    pub contexts: Vec<u16>,
+    pub ce32s: Vec<u32>,
+    // TOML integers are signed 64-bit, so the range of u64 isn't available
+    pub ces: Vec<i64>,
+    pub last_primaries: Vec<u16>, // length always supposed to be 4
+}
+
+#[derive(serde::Deserialize)]
+pub struct CollationDiacritics {
+    pub ce32s: Vec<u32>,
+}
+
+#[derive(serde::Deserialize)]
+pub struct CollationJamo {
+    pub ce32s: Vec<u32>,
+}
+
+#[derive(serde::Deserialize)]
+pub struct CollationMetadata {
+    pub bits: u32,
+}
+
+#[derive(serde::Deserialize)]
+pub struct CollationReordering {
+    pub min_high_no_reorder: u32,
+    pub reorder_table: Vec<u8>,
+    pub reorder_ranges: Vec<u32>,
+}
+
+macro_rules! collation_provider {
+    ($marker:ident, $provider:ident, $key:ident, $serde_struct:ident, $suffix:literal, $conversion:expr, $toml_data:ident) => {
+        use icu_collator::provider::$marker;
+
+        pub struct $provider {
+            data: HashMap<String, $serde_struct>,
+        }
+
+        impl TryFrom<&DatagenOptions> for $provider {
+            type Error = DataError;
+            fn try_from(options: &DatagenOptions) -> Result<Self, Self::Error> {
+                Ok(Self::try_new(&options.collator_data_root.as_ref().ok_or_else(|| DataError::custom("collator data root required"))?)?)
+            }
+        }
+
+        /// A data provider reading from .toml files produced by the ICU4C genrb tool.
+        impl $provider {
+            pub fn try_new(root_dir: &Path) -> Result<Self, DataError> {
+                let data = $provider::load_data(root_dir).map_err(|e| {
```

Comment:
This is now refactored. Beyond making this run, I think it's the best to get this landed and tweak after, because otherwise another datagen refactoring always blocks this.